minvws/nl-kat-coordination
# Improve the KATalogus `/plugins` endpoint performance

### Changes

And add a bulk endpoint for the declarations of the CSV upload. The main fixes in the KATalogus are:
- Do not query the database per boefje id on each request to fetch the `enabled` state, just do 1 query per request
- Query both boefje and normalizers at the same time
- Do not query the existence of the organization on each request: this is added as a filter parameter anyways and this should be checked in Rocky
- Cache boefje/normalizer module loading not on an instance-level but with a function @cache decorator, since then we have the cache available for other requests as well where before we had a new instance of the local repository per request.

### Issue link

Closes https://github.com/minvws/nl-kat-coordination/issues/3903
Closes https://github.com/minvws/nl-kat-coordination/issues/2524

### Demo

#### Before

Snapshot of `htop` (also notice the postgres process because we did 100+ queries per second, because we managed to handle 2 requests to the `/plugins` endpoint per second):
<img width="1511" alt="Scherm­afbeelding 2024-12-02 om 09 54 43" src="https://github.com/user-attachments/assets/222d3aee-ef7e-42db-b88c-ac164623daa2">

Requests per second (RPS) of the `plugin` endpoint, tested using [Locust](https://locust.io/):
![Scherm­afbeelding 2024-11-29 om 11 34 50](https://github.com/user-attachments/assets/971eac40-8b4e-41f6-9021-b76260c94b1c)


#### After
Requests per second (RPS) of the `plugin` endpoint, tested using [Locust](https://locust.io/):
![Scherm­afbeelding 2024-11-29 om 11 32 31](https://github.com/user-attachments/assets/cf15828d-a3bc-48b6-b625-4f238e2f3709)

Snapshot of `htop` (the boefje api processes where nowhere near the top):
<img width="1233" alt="Scherm­afbeelding 2024-12-02 om 08 30 43" src="https://github.com/user-attachments/assets/60cda178-a409-4ac0-b8b5-63145ef27d59">


### QA notes

Please verify that the CSV upload still works and the scheduler picks up tasks, also from duplicated boefjes! Perhaps use the [test server](https://kattest03.openkat.dev/) that I used to verify the interface works OK with 300 organizations.

---

### Code Checklist





---

## Checklist for code reviewers:

Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_code.md) into your comment.

---

## Checklist for QA:

Copy-paste the checklist from [the docs/source/templates folder](https://github.com/minvws/nl-kat-coordination/blob/main/docs/source/templates/pull_request_template_review_qa.md) into your comment.
